### PR TITLE
W-13817421 Schema Definition

### PR DESCRIPTION
**Pull request by glenn-rodgers-sf**
# Writer's Quality Checklist

Before merging your PR, did you:

- [ ] Run spell checker
- [ ] Run link checker to check for broken xrefs
- [ ] Check for orphan files
- [ ] Perform a local build and do a final visual check of your content, including checking for:
  - Broken images
  - Dead links
  - Correct rendering of partials if they are used in your content
  - Formatting issues, such as:
    - Misnumbered ordered lists (steps) or incorrectly nested unordered lists
    - Messed up tables
    - Proper indentation
    - Correct header levels
- [ ] Receive final review and signoff from:
  - Technical SME
  - Product Manager
  - Editor or peer reviewer
  - Reporter, if this content is in response to a reported issue (internal or external feedback)
- [ ] If applicable, verify that the software actually got released

### CODE REVIEW

File: gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc [hunk 1]
```diff
@@ -1,5 +1,295 @@
...
+You define properities as required by listing them in the `spec.required` object. For example, the following policy requires a `password` but not `username`:
+
+[source,yaml]
+----
+properties:
+    username:
+        type: string
+    password:
+        type: string
+Required:
```
Comment (JaatadiaMulesoft) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc]: lowercase on the 'required'

File: gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc [hunk 2]
```diff
@@ -1,5 +1,295 @@
...
+[source,yaml]
+----
+properties:
+    username:
+        type: string
+        default: user1
+    password:
+        type: string
+        default: 12345678
+Required:
```
Comment (JaatadiaMulesoft) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc]: idem

File: gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc [hunk 3]
```diff
@@ -1,5 +1,295 @@
...
++
+The supported formating types are:
++
+* `uri`: Specifies the expected text is a URI
+* `ip-range`: Specifies the expected text is a range or IPs
+* `dataweave`: specifies the expected text is a dataweave variable
++
+Defining a string as the `uri` or `ip-range` provides additional validations, but defining a string as `dataweave` notifies Flex Gateway that the input must be transformed. For more information about supported Dataweave transformations, see xref:policies-flex-dataweave-support.adoc[].
++
+The following yaml snippit demonstrates all string formats:
```
Comment (JaatadiaMulesoft) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc]: snippet

**Review by JaatadiaMulesoft — APPROVED** [review]

File: gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc [hunk 4]
```diff
@@ -1,5 +1,362 @@
-= Creating a Schema Definition
+= Defining a Policy Schema Definition
 ifndef::env-site,env-github[]
 include::_attributes.adoc[]
 endif::[]
-:imagesdir: ../assets/images
\ No newline at end of file
+:imagesdir: ../assets/images
+
+Similar to the policies listed in the xref:policies-included-directory.adoc[], custom policies also have configuration paramaters. Configuration parameters enable you to improve the usability of you custom policy and support a wider range of API configurations.
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc]:
```suggestion
Similar to the policies listed in the xref:policies-included-directory.adoc[], custom policies have configuration parameters. Use the configuration parameters to improve the usability of your custom policy and support a wider range of API configurations.
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc [hunk 5]
```diff
@@ -1,5 +1,362 @@
...
+
+You define your own custom configuration parameters by editing the `definition/gcl.yaml` file. The `gcl.yaml` file is written in Gateway Configuration Language (GCL). However, you do not need prior GCL knowledge to use this tutorial.
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc]:
```suggestion
To define custom configuration parameters, edit the `definition/gcl.yaml` file. The `gcl.yaml` file is written in Gateway Configuration Language (GCL). However, you do not need prior GCL knowledge to use this tutorial.
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc [hunk 6]
```diff
@@ -1,5 +1,362 @@
...
+
+The configuration parameters you define in the `gcl.yaml` appear as variables in the `src/generated/config.rs` file after you run the `build-asset-file` command. You set any user input you need for writing the rust source code in the `gcl.yaml` file.
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc]:
```suggestion
The configuration parameters you define in the `gcl.yaml` appear as variables in the `src/generated/config.rs` file after you run the `build-asset-file` command. You set any user input you need for writing the Rust source code in the `gcl.yaml` file.
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc [hunk 7]
```diff
@@ -1,5 +1,362 @@
...
+
+Additionally, the configuration parameters appear in the API Manager UI when applying a policy to Flex Gateway in Connected Mode. When applying the policy to Flex Gateway running in Local Mode, it is your responsibility to distribute a configuration `yaml` file for your policy.
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc]:
```suggestion
Configuration parameters appear in the API Manager UI when applying a policy to Flex Gateway running in Connected Mode. When applying the policy to Flex Gateway running in Local Mode, it is your responsibility to distribute a configuration `yaml` file for your policy.
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc [hunk 8]
```diff
@@ -1,5 +1,362 @@
...
+
+The default `gcl.yaml` file appears as the following:
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc]:
```suggestion
The default `gcl.yaml` file has the following content:
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc [hunk 9]
```diff
@@ -1,5 +1,362 @@
...
+
+To define a schema definition:
+
+. <<define-metadata>>
+. <<define-parameters>>
+. <<use-custom-parameters-in-rust-source-code>>
+
+== Define Metadata
+
+By adding labels in the metadata section, you can enrich the data shown in the UI when applying the policy.
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc]:
```suggestion
Add labels in the metadata section to enrich the data shown in the UI when applying the policy.
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc [hunk 10]
```diff
@@ -1,5 +1,362 @@
...
+
+To define the metadata, edit the `metadata` in the `gcl.yaml` file.
+
+The supported labels are:
+
+* `name`: The policy name that appears in the UI
+* `description`: A brief explanation of the policy's behavior
+* `category`: The category the policy appears under. Supported values are: `security`, `compliance`, `transformation`, `quality of service`, and `troubleshooting`.
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc]:
```suggestion
* `name`: The policy name that appears in the UI.
* `description`: A brief explanation of the policy's behavior.
* `category`: The category the policy appears under. Supported values are: `security`, `compliance`, `transformation`, `quality of service`, and `troubleshooting`. 
```
Adjusted punctuation to be consistent with the last item

File: gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc [hunk 11]
```diff
@@ -1,5 +1,362 @@
...
+        title: Json Threat Protection
+        description: Protects against malicious JSON in API requests.
+        category: Security
+----
+
+`labels` are suggested but not required.
+
+== Define Parameters
+
+To add parameters, edit `spec.properties` section  in the `gcl.yaml` file.
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc]:
```suggestion
To add parameters, edit the `spec.properties` section in the `gcl.yaml` file.
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc [hunk 12]
```diff
@@ -1,5 +1,362 @@
...
+----
+
+After defining your properties, you can:
+
+* <<define-required-properties-and-default-values>>
+* <<define-supported-types>>
+
+=== Define Required Properties and Default Values
+
+For each property in the `gcl.yaml` file, you can define the property as `required` and set a `default` value.
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc]:
```suggestion
You can define each property listed in the `gcl.yaml` file as `required` and set a `default` value for the property.
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc [hunk 13]
```diff
@@ -1,5 +1,362 @@
...
+
+You define properities as required by listing them in the `spec.required` object. For example, the following policy requires a `password` but not `username`:
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc]:
```suggestion
To define properties as required, list them in the `spec.required` object. For example, the following policy requires a `password` but not a `username`:
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc [hunk 14]
```diff
@@ -1,5 +1,362 @@
...
+        type: string
+    password:
+        type: string
+required:
+    - password
+----
+
+In this configuration example, configuring a policy without a password results in a failed configuration.
+
+You can also define a default value for a property. For Example:
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc]:
```suggestion
The following example defines a default value for each property:
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc [hunk 15]
```diff
@@ -1,5 +1,362 @@
...
+        type: string
+        default: user1
+    password:
+        type: string
+        default: 12345678
+required:
+    - password
+----
+
+Defining a default value for a required property means that if the field is left blank when configuring the policy, the parameter is set to the default value. Defining a default value for a parameter that isn't required, means that the default value appears as a suggested value in the UI. If a non-required parameter with a default value is left blank during configuration, there is no configured value for the parameters. In the following configuration table, if the policy was applied in API Manager, `user1` would appear as a suggested value for `username` but does not affect if the value is configured:
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc]:
```suggestion
If you don't set a value for a required property when configuring a policy, the property uses the default value. The UI suggests the configured default value for properties that are not required. If you don't specify the value for non-required properties, the property does not use the default value. 

In the following configuration table, if the policy was applied in API Manager, `user2` would appear as a suggested value for `username` but does not affect if the value is configured:
```
I tried to make this sentence a bit more clear and less wordy, please check if it is Ok before applying the suggestion. I also changed the value from user1 to user2 to match the example, not sure if that's right.

File: gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc [hunk 16]
```diff
@@ -1,5 +1,362 @@
...
+"username": "user2"}` | Success
+| `{"password": "password"}` | `{"password": "password"}` | Success
+| `{"password": "password", +
+ "username": "user2"}` | `{"password": "password", +
+"username": "user2"}` | Success
+|===
+
+=== Define Supported Types
+
+You can define a supported data type for each policy parameter to ensure the policy receives the correct data.
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc]:
```suggestion
Define a supported data type for each policy parameter to ensure the policy receives the correct data.
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc [hunk 17]
```diff
@@ -1,5 +1,362 @@
...
+          type: string
+      required:
+        - id
+        - name
+      default: {"id": "1", "name": "John Doe"}
+----
+
+== Use Custom Parameters in Rust Source Code
+
+After you modify the `gcl.yaml` file you can propagate the changes to the `src/generated/config.rs` rust file and the json schema for the UI by running the `build-asset-files` command:
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc]:
```suggestion
After you modify the `gcl.yaml` file, run the `build-asset-files` command to propagate the changes to the `src/generated/config.rs` rust file and the JSON schema for the UI:
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc [hunk 18]
```diff
@@ -1,5 +1,362 @@
...
+
+[source,cmd]
+----
+make build-asset-files
+----
+
+Additionally compiling the project with the `build` command also propagates the changes:
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc]:
```suggestion
Additionally, compiling the project with the `build` command also propagates the changes:
```

File: gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc [hunk 19]
```diff
@@ -1,5 +1,362 @@
...
+
+[source,cmd]
+----
+make build
+----
+
+For more information about compiling your policy, see xref:policies-pdk-compile-policies.adoc[].
+
+Properties defined in the `gcl.yaml` files are mapped to the rust data types in the `config.rs` rust file as follows:
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc]:
```suggestion
Properties defined in the `gcl.yaml` files are mapped to the Rust data types in the `config.rs` Rust file as follows:
```
I'm marking Rust as uppercase assuming this is talking about the programming language. Please confirm if this is correct.

File: gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc [hunk 20]
```diff
@@ -1,5 +1,362 @@
...
+[IMPORTANT] 
+----
+If a parameter in `gcl.yaml` is not defined as ‘required’, the parameters counterpart in the `config.rs` is wrapped in an ‘Option’ object.
+
+Validate the the parameter is present before unwrapping. Unwrapping a parameter that is not present may cause a `panic!`.
+
+You can use one of the following validation examples to ensure the parameter is present:
+
+* `if let Some(val) = myValue`
+* `if myValue.is_ome() {myValue.unwrap()}`
```
Comment (Cristian-Venticinque) [inline gateway/1.5/modules/ROOT/pages/policies-pdk-create-schema-definition.adoc]:
```suggestion
If a parameter in `gcl.yaml` is not defined as ‘required’, the parameter's counterpart in the `config.rs` is wrapped in an ‘Option’ object.

Validate that the parameter is present before unwrapping. Unwrapping a parameter that is not present may cause a `panic!`.

Use one of the following validation examples to ensure the parameter is present:

* `if let Some(val) = myValue`
* `if myValue.is_ome() {myValue.unwrap()}`
```

**Review by Cristian-Venticinque — APPROVED** [review]
Approved with comments.

**Review by Cristian-Venticinque — APPROVED** [review]

**Pull request merged**